app/database/permission_granted_store_compute_all_access.go

```diff
@@ -0,0 +1,146 @@
+	s.mustBeInTransaction()
+
+	// ------------------------------------------------------------------------------------
+	// Here we declare and prepare DB statements that will be used by the function later on
+	// ------------------------------------------------------------------------------------
+	var stmtMarkChildrenOfChildrenAsSelf, stmtDeleteProcessedChildren, stmtUpdatePermissionsGenerated, stmtMarkSelfAsChildren *sql.Stmt
+	var err error
+
+	// marking group-item pairs whose parents are marked with propagate_access='children' as 'self'
+	const queryMarkChildrenOfChildrenAsSelf = `
+		INSERT INTO permissions_propagate (group_id, item_id, propagate_access)
+		SELECT
+			parents.group_id,
+			items_items.child_item_id,
+			'self' as propagate_access
+		FROM items_items
+		JOIN permissions_generated AS parents
+			ON parents.item_id = items_items.parent_item_id
+		JOIN permissions_propagate AS parents_propagate
+			ON parents_propagate.group_id = parents.group_id AND parents_propagate.item_id = parents.item_id AND
+				 parents_propagate.propagate_access = 'children'
+		ON DUPLICATE KEY UPDATE propagate_access='self'`
+	stmtMarkChildrenOfChildrenAsSelf, err = s.db.CommonDB().Prepare(queryMarkChildrenOfChildrenAsSelf)
+	mustNotBeError(err)
+	defer func() { mustNotBeError(stmtMarkChildrenOfChildrenAsSelf.Close()) }()
+
+	// deleting 'children' groups_items_propagate
+	const queryDeleteProcessedChildren = `DELETE FROM permissions_propagate WHERE propagate_access = 'children'`
+	stmtDeleteProcessedChildren, err = s.db.CommonDB().Prepare(queryDeleteProcessedChildren)
+	mustNotBeError(err)
+	defer func() { mustNotBeError(stmtDeleteProcessedChildren.Close()) }()
+
+	// computation for group-item pairs marked as 'self' in permissions_propagate (so normally all of them)
+	const queryUpdatePermissionsGenerated = `
+		INSERT INTO permissions_generated
+			(group_id, item_id, can_view_generated, can_grant_view_generated, can_watch_generated, can_edit_generated, is_owner_generated)
+		SELECT
+			permissions_propagate.group_id,
+			permissions_propagate.item_id,
+			IF(granted.is_owner, 'solution', GREATEST(
+				IFNULL(granted.can_view_value, 1),
+				IFNULL(new_data.can_view_value, 1)
+			)) AS can_view_generated,
+			IF(granted.is_owner, 'transfer', GREATEST(
+				IFNULL(granted.can_grant_view_value, 1),
+				IFNULL(new_data.can_grant_view_value, 1)
+			)) AS can_grant_view_generated,
+			IF(granted.is_owner, 'transfer', GREATEST(
+				IFNULL(granted.can_watch_value, 1),
+				IFNULL(new_data.can_watch_value, 1)
+			)) AS can_watch_generated,
+			IF(granted.is_owner, 'transfer', GREATEST(
+				IFNULL(granted.can_edit_value, 1),
+				IFNULL(new_data.can_edit_value, 1)
+			)) AS can_edit_generated,
+			IFNULL(granted.is_owner, 0) AS is_owner_generated
+		FROM permissions_propagate
+		LEFT JOIN LATERAL (
+			SELECT
+				MAX(LEAST(
+					IF(LEAST(items_items.descendants_and_solution_view_propagation_value+2, parent.can_view_generated_value) = 3,
```

Comment:
Wow arithmetic on enum value becomes very difficult to read, especially combined with LEAST and IFs ;-). Is there a reason not to the pure logic (syntax verified):
```sql
        MAX(
          CASE parent.can_view_generated 
          WHEN 'none', 'info' THEN 1 -- 'none'
          WHEN 'content' THEN 
            CASE items_items.content_view_propagation
            WHEN 'as_info' THEN 2 -- 'info'
            WHEN 'as_content' THEN 3 -- 'content'
            ELSE 1 -- 'none' 
            END
          ELSE 
            CASE items_items.descendants_and_solution_view_propagation
            WHEN 'descendants' THEN 4 --'content_with_descendants'
            WHEN 'descendants_and_solution' THEN parent.can_view_generated_value
            ELSE 
              CASE items_items.content_view_propagation_value
              WHEN 'as_info' THEN 2 -- 'info'
              WHEN 'as_content' THEN 3 -- 'content'
              ELSE 1 --'none'
              END
            END
          END) AS can_view_value,
```
OR 
```sql
        MAX(
          CASE 
          WHEN parent.can_view_generated IN ('none', 'info') THEN 1 -- none
          WHEN parent.can_view_generated = 'content' OR items_items.descendants_and_solution_view_propagation_value = 'none' THEN 
            CASE items_items.content_view_propagation_value
            WHEN 'as_info' THEN 2 -- info
            WHEN 'as_content' THEN 3 -- content
            ELSE 2 -- none
            END
          WHEN items_items.descendants_and_solution_view_propagation = 'descendants' THEN 4 -- content_with_descendants
          ELSE parent.can_view_generated_value 
          END) AS can_view_value,
```
It is so much more readable, verifiable, modifiable.

Reply:
done